File: qvirtservice/src/VirtualizationService.cpp
/*
  * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
  * SPDX-License-Identifier: BSD-3-Clause-Clear
*/

#include "VirtualizationService.h"
#include<thread>
#include <poll.h>
#include <fstream>
#include <json/json.h>
#include <cutils/uevent.h>
#include <cutils/properties.h>
#include <android-base/properties.h>

using android::base::WaitForProperty;
using aidl::vendor::qti::qvirt::VirtualMachine;

namespace aidl {
namespace vendor {
namespace qti {
namespace qvirt {

static void parse_event(const char *msg, struct uevent *uevent) {
    uevent->vm_name = "";
    uevent->EVENT = "";
    while (*msg) {
       if (!strncmp(msg, "EVENT=", 6)) {
           msg += 6;
           uevent->EVENT = msg;
       } else if (!strncmp(msg, "vm_name=", 8)) {
           msg += 8;
           uevent->vm_name = msg;
       }
       // advance to after the next \0
       while (*msg++);
    }
}

static void handle_uevent_event(std::map<std::string,
        std::shared_ptr<VirtualMachine> >& vmObject_map) {

    char msg[UEVENT_MSG_LEN + 2];
    int n;
    struct uevent uevent;

    while ((n = uevent_kernel_multicast_recv(uevent_fd, msg, UEVENT_MSG_LEN)) > 0) {
        if (n >= UEVENT_MSG_LEN) /* overflow -- discard */
            continue;

        msg[n] = '\0';
        msg[n + 1] = '\0';
        parse_event(msg, &uevent);
        if (uevent.EVENT != NULL && strlen(uevent.EVENT) != 0
                    && uevent.vm_name != NULL && strlen(uevent.vm_name) != 0)
        {
            VM_LOGI("uevent received is: EVENT = %s , vm_name = %s",
                    uevent.EVENT, uevent.vm_name);
            std::unique_lock<std::mutex> lk(mMutex_);
            if (vmObject_map.find(uevent.vm_name) == vmObject_map.end())
            {
                VM_LOGE("Invalid vm_name received from uevent");
                lk.unlock();
                continue;
            }
            std::shared_ptr<VirtualMachine> vmObject = vmObject_map[uevent.vm_name];
            lk.unlock();
            VM_LOGV("handle_uevent_event: Initiating request to notify state change to %s clients",
                    uevent.vm_name);
            if (vmObject != nullptr) {
                vmObject->notify_clients_locked(uevent.EVENT);
            }
        }
    }

}

static void uevent_listener(std::map<std::string,std::shared_ptr<VirtualMachine> >& vmObject_map){

    VM_LOGI("started uevent listener thread\n");
    uevent_fd = uevent_open_socket(64 * 1024, true);

    if (uevent_fd < 0) {
        VM_LOGE("uevent_open_socket() failed\n");
        return;
    }
    fcntl(uevent_fd, F_SETFL, O_NONBLOCK);
    pollfd ufd;
    ufd.events = POLLIN;
    ufd.fd = uevent_fd;

    while (true) {
        ufd.revents = 0;
        int nr = poll(&ufd, 1, -1);

        if (nr <= 0) {
            VM_LOGE("poll() of uevent socket failed, continuing");
            continue;
        }
        if (ufd.revents & POLLIN) {
            handle_uevent_event(vmObject_map);
        }
    }
}

int VirtualizationService::parseVMConfigFile() {

    string vendorConfigFile(VENDOR_CONFIG_FILE);

    // Attempt to open vendor-specific config file
    ifstream inputFile(vendorConfigFile, ifstream::in);

    if (!inputFile.is_open())
    {
        VM_LOGE("Couldn't open vendor vm config file (%s), aborting.", vendorConfigFile.c_str());
        return -1;
    }
    VM_LOGI("Using '%s' vm config file", vendorConfigFile.c_str());

    Json::CharReaderBuilder reader;
    Json::Value root;
    std::string err;

    reader["collectComments"] = false;
    if (!Json::parseFromStream(reader, inputFile, &root, &err))
    {
        VM_LOGE("Parsing configuration file '%s' failed with error: %s", vendorConfigFile.c_str(),
                err.c_str());

       // Close config file
       inputFile.close();
       return -1;
    }

    // Close config file
    inputFile.close();

    // Read root node
    const Json::Value& jsonVmConfig = root["qvirtmgr"];

    const Json::Value& jsonConfigArray = jsonVmConfig["vm_config"];
    if (!jsonVmConfig || !jsonConfigArray)
    {
        VM_LOGE("VM configurations found in '%s' is invalid", vendorConfigFile.c_str());
        return -1;
    }

    for (const Json::Value& jsonConfig : jsonConfigArray)
    {
        if (!jsonConfig.isMember("name") || jsonConfig["name"].isNull())
        {
            VM_LOGE("Missing value for vmName,skipping entry");
            continue;
        }

        if (!jsonConfig.isMember("enable") || !jsonConfig["enable"].isBool())
        {
            VM_LOGE("Missing or Invalid value for enable, skipping entry for %s",
                    jsonConfig["name"].asString().c_str());
            continue;
        }

        VirtualMachine::VmParameters vmParameters(jsonConfig["name"].asString(),
                jsonConfig["enable"].asBool());

        if (jsonConfig.isMember("autostart") && jsonConfig["autostart"].isBool())
        {
            vmParameters.autostart = jsonConfig["autostart"].asBool();
        }
        else
        {
            VM_LOGE("Missing or Invalid value for autostart, skipping entry for %s",
                    jsonConfig["name"].asString().c_str());
            continue;
        }

        if (jsonConfig.isMember("on_demand_start_supported") &&
                jsonConfig["on_demand_start_supported"].isBool())
        {
            vmParameters.on_demand_start_supported =
                    jsonConfig["on_demand_start_supported"].asBool();
        }

        if (jsonConfig.isMember("no_fs_dependency") && jsonConfig["no_fs_dependency"].isBool())
        {
            vmParameters.noFSDependency_ = jsonConfig["no_fs_dependency"].asBool();
        }
        else
        {
            VM_LOGE("Missing or Invalid value for no_fs_dependency, skipping entry for %s",
                    jsonConfig["name"].asString().c_str());
            continue;
        }

        if (jsonConfig.isMember("disk"))
        {
            // Read disk properties
            const Json::Value& jsonDiskArray = jsonConfig["disk"];

            bool diskparams = true;
            for (const Json::Value& jsonDisk : jsonDiskArray)
            {
                if (!jsonDisk.isMember("image") || !jsonDisk.isMember("label") ||
                        !jsonDisk.isMember("read_write") || !jsonDisk["read_write"].isBool() )
                {
                    diskparams = false;
                    break;
                }
                string image = jsonDisk["image"].asString();
                string label = jsonDisk["label"].asString();
                bool readWrite = jsonDisk["read_write"].asBool();

                vmParameters.disk_.emplace_back(image, label, readWrite);
            }
            if (!diskparams)
            {
                VM_LOGE("Missing or Invalid disk parameters, skipping entry for %s",
                        jsonConfig["name"].asString().c_str());
                continue;
            }
        }

        if (jsonConfig.isMember("try_count") && jsonConfig["try_count"].isUInt())
        {
            vmParameters.bootTryCount_ = jsonConfig["try_count"].asUInt();
        }

        if (jsonConfig.isMember("boot_wait_time") && jsonConfig["boot_wait_time"].isUInt())
        {
            vmParameters.bootWaitTime_ = jsonConfig["boot_wait_time"].asUInt();
        }
        else
        {
            VM_LOGE("Missing or Invalid value for boot_wait_time, skipping entry for %s",
                    jsonConfig["name"].asString().c_str());
            continue;
        }

        if (jsonConfig.isMember("boot_complete_timeout") &&
                jsonConfig["boot_complete_timeout"].isUInt())
        {
            vmParameters.bootCompleteTimeout_ = jsonConfig["boot_complete_timeout"].asUInt();
        }

        string temp("");

        if (jsonConfig.isMember("boot_ops"))
        {
            temp = jsonConfig["boot_ops"].asString();
            vmParameters.bootOperation_ = VirtualMachine::getControlState(temp);
            vmParameters.vmBinary_ = string(VM_BINARY_FILE);
        }
        else
        {
            VM_LOGE("Missing or Invalid value for boot_ops, skipping entry for %s",
                    jsonConfig["name"].asString().c_str());
            continue;
        }

        vmParameters_map.insert({vmParameters.vmName_, vmParameters});
    }
    VM_LOGI("Number of vm configurations = %lu", vmParameters_map.size());
    return 0;
}

std::shared_ptr<VirtualMachine> VirtualizationService::getVm_Internal(string in_vmName) {

    if (vmObject_map.find(in_vmName) != vmObject_map.end()) {
        return vmObject_map[in_vmName];
    }

    VM_LOGI("getVm_Internal: Creating vmObject for %s", in_vmName.c_str());
    std::shared_ptr<VirtualMachine> mVirtualMachine = SharedRefBase::make<VirtualMachine>();

    if (mVirtualMachine != nullptr) {
        mVirtualMachine->vmInstance.vmName_ = in_vmName;
        mVirtualMachine->verbose = verbose;
        mVirtualMachine->setVmStatusProperty("NOT_STARTED");
        mVirtualMachine->vmInstance.vmParameters_ = vmParameters_map[in_vmName];
        vmObject_map.insert({in_vmName, mVirtualMachine});
        VM_LOGV("size of vmObject map is %d", vmObject_map.size());
    }
    return mVirtualMachine;
}

static void autostarthandler(std::shared_ptr<VirtualMachine> virtualMachineObj)
{
    VM_LOGI("autostarthandler: Launching %s", virtualMachineObj->vmInstance.vmName_.c_str());
    virtualMachineObj->launch_autostartVMs();
    VM_LOGV("Exiting from autostarthandler for vm %s",
            virtualMachineObj->vmInstance.vmName_.c_str());
}

VirtualizationService::VirtualizationService(int argc, char **argv) {

    if (argc>1 && strcmp(argv[1],"-v") == 0)
    {
        verbose = true;
    }
    std::thread thread(uevent_listener, std::ref(vmObject_map));
    thread.detach();
    int state = parseVMConfigFile();
    if (state != 0)
    {
        exit(EXIT_FAILURE);
    }
    else
    {
        VM_LOGI("parsing vm config file success");
    }
    vector<std::thread> vmThreads_noFSDependent;
    vector<std::thread> vmThreads_FSDependent;

    for (auto i : vmParameters_map)
    {
        if (i.second.isEnabled_)
        {
            VM_LOGV("autoboot enabled for %s", i.first.c_str());

            std::unique_lock<std::mutex> lk(mMutex_);
            getVm_Internal(i.first);

            if (vmObject_map.find(i.first) == vmObject_map.end()) {
                VM_LOGE("vmObject not found for %s", i.first.c_str());
                continue;
            }

            if (i.second.autostart)
            {
                if (i.second.noFSDependency_) {
                    vmThreads_noFSDependent.push_back(std::thread(autostarthandler,
                            vmObject_map[i.first]));
                }
                else {
                    vmThreads_FSDependent.push_back(std::thread(autostarthandler,
                            vmObject_map[i.first]));
                }
            }

            lk.unlock();
        }
        else {
            char vmStatusProp[BUFFER_MAX];
            snprintf(vmStatusProp, BUFFER_MAX, VM_STATUS_PROPERTY, i.first.c_str());
            android::base::SetProperty(vmStatusProp, "NOT_STARTED");
        }
    }

    for (int i = 0; i < vmThreads_FSDependent.size(); i++)
    {
        vmThreads_FSDependent[i].detach();
    }

    for (int i = 0; i < vmThreads_noFSDependent.size(); i++)
    {
        vmThreads_noFSDependent[i].join();
    }
}

ndk::ScopedAStatus VirtualizationService::getVm(const std::string& in_vmName,
        std::shared_ptr<IVirtualMachine>* iVirtualMachine) {

    VM_LOGI("getVm: Requested vm handle for %s from pid = %ld", in_vmName.c_str(),
            AIBinder_getCallingPid());

    std::unique_lock<std::mutex> lk(mMutex_);

    if (vmParameters_map.find(in_vmName) == vmParameters_map.end())
    {
        VM_LOGE("getVm: Invalid vmName argument passed, rejecting request");
        return ndk::ScopedAStatus::fromExceptionCode(EX_ILLEGAL_ARGUMENT);
    }
    else if (!vmParameters_map[in_vmName].isEnabled_)
    {
        VM_LOGE("getVm: enabled bit is false for %s, rejecting request", in_vmName.c_str());
        return ndk::ScopedAStatus::fromExceptionCode(EX_UNSUPPORTED_OPERATION);
    }

    *iVirtualMachine = getVm_Internal(in_vmName);

    if (iVirtualMachine == nullptr)
    {
        VM_LOGE("getVm: VirtualMachineObject is null, rejecting request");
        return ndk::ScopedAStatus::fromExceptionCode(EX_NULL_POINTER);
    }

    return ndk::ScopedAStatus::ok();

}

}
}
}
}

File: qvirtservice/rs_src/virtualization_service.rs
/*
 * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
 * SPDX-License-Identifier: BSD-3-Clause-Clear
*/

use crate::virtual_machine::{VirtualMachine, VmInstance, VmParameters, UeventInfo};

use crate::utils::UEvent;

use rustutils::system_properties;

use nix::fcntl::{fcntl, FcntlArg, OFlag};
use nix::poll::{poll, PollFd, PollFlags, PollTimeout};
use std::os::unix::io::RawFd;
use std::{
    collections::HashMap,
    error::Error,
    fs::File,
    sync::{Arc, Mutex},
    thread,
    time::Duration,
};

use libc::_exit;

use log::{debug, error, info};

use serde_json::Value;

use std::os::fd::{BorrowedFd};

use vendor_qti_qvirt::aidl::vendor::qti::qvirt::{
    IVirtualMachine::IVirtualMachine,
    IVirtualizationService::{
        BnVirtualizationService, BpVirtualizationService,
        IVirtualizationService,
    },
};
use vendor_qti_qvirt::binder::{
    BinderFeatures, ExceptionCode, Interface, Status, Strong, ThreadState,
};

//use vendor_qti_qvirtvendor::aidl::vendor::qti::qvirtvendor::IVendorVM::IVendorVM;

static VENDOR_CONFIG_FILE: &str = "/vendor/etc/qvirtmgr-vndr.json";
static BOOT_COMPLETE_PROP: &str = "sys.boot_completed";

pub struct VmInstanceWrapper {
    // VmInstance
    // - Arc -> Shared between threads (here, autostart, getVm).
    // - Mutex -> Needs to be mutable anywhere (autostart_complete, etc.)
    instance: Arc<Mutex<VmInstance>>,
    enabled: bool,
    boot_complete_timeout: u16,
    enabled_socs: Vec<String>,
    current_soc: String,
}

//                          < name , VmInstanceWrapper>
type VmInstanceMap = HashMap<String, VmInstanceWrapper>;

pub struct VirtualizationService {
    // VmInstanceMap
    // - Arc -> Shared between threads (here, uevent).
    // - !Mutex -> Only modified in here once in constructor.
    vm_instance_map: Arc<VmInstanceMap>,
}

impl VirtualizationService {
    pub fn to_binder(self) -> Strong<dyn IVirtualizationService> {
        BnVirtualizationService::new_binder(self, BinderFeatures::default())
    }

    pub fn get_descriptor() -> String {
        BpVirtualizationService::get_descriptor().to_string()
    }

    pub fn virtualization_service() -> Self {
        let mut vm_instance_map: VmInstanceMap = Default::default();

        // Store the vm's thread handle for the no_fs_dependent ones.
        let mut no_fs_dependent_handles = Vec::new();

        // Store autostart vms
        let mut autostart_vms = Vec::<(String, bool)>::new();

        // Parse the config file
        if let Ok(vm_parameters_list) = Self::parse_vm_config_file() {
            info!("Parsing VM Config File Success.");

            for vm_param in vm_parameters_list {
                // Save data now for easy access (before locked in mutex).
                let name = vm_param.name.clone();
                let enabled = vm_param.enable.clone();
                let boot_complete_timeout =
                    vm_param.boot_complete_timeout.clone();
                let enabled_socs = vm_param.enabled_socs.clone();
                let mut current_soc = String::new();
                if !(enabled_socs.is_empty()) {
                    if let Ok(Some(value)) = system_properties::read("ro.boot.product.vendor.sku") {
                        debug!("Current sku value is {}",value);
                        current_soc = value.clone();
                        if (vm_param.autostart) && enabled_socs.contains(&value) {
                            autostart_vms.push((
                                name.clone(),
                                vm_param.no_fs_dependency.clone(),
                            ));
                        }
                    }
                }
                else if vm_param.autostart && vm_param.enable {
                    autostart_vms.push((
                        name.clone(),
                        vm_param.no_fs_dependency.clone(),
                    ));
                }

                // Create a VmInstance and put the vm_params inside.
                // When getVm called, give out a VirtualMachine wrapping a ref to the instance
                let wrpr = VmInstanceWrapper {
                    instance: Arc::new(Mutex::new(VmInstance::new(vm_param))),
                    enabled: enabled,
                    boot_complete_timeout: boot_complete_timeout,
                    enabled_socs: enabled_socs,
                    current_soc: current_soc,
                };

                vm_instance_map.insert(name, wrpr); // Put a ref count in the map
            }
        } else {
            unsafe {
                _exit(1);
            }
        }

        let service = VirtualizationService {
            vm_instance_map: Arc::new(vm_instance_map),
        };

        // Create the uevent thread
        let vm_instance_map_clone = service.vm_instance_map.clone();
        thread::spawn(move || {
            Self::uevent_listener(vm_instance_map_clone);
        });

        // Launch autostart VMs
        for (name, no_fs_dependency) in autostart_vms {
            let wrapper = service.vm_instance_map.get(&name).unwrap();
            let timeout_for_thread = wrapper.boot_complete_timeout.clone();
            let instance_for_thread = wrapper.instance.clone();
            if no_fs_dependency {
                let handle = thread::spawn(move || {
                    let mut vm_ssr_enablecheck:bool = false;
                    let mut vm_autostart_done:bool = false;
                    if let Ok(mut vm) = instance_for_thread.lock() {
                        vm.launch_autostart_vm();
                        vm_ssr_enablecheck = vm.vm_parameters.vm_ssr_enable;
                        vm_autostart_done = vm.autostart_done;
                        drop(vm);
                    }
                    if vm_ssr_enablecheck && vm_autostart_done{
                        if let Ok(mut vm) = instance_for_thread.lock() {
                            match vm.autostart_connectvm() {
                                Ok(0)=>{
                                    info!("VM userpspace connection is successful");
                                },
                                Ok(_)=>{
                                    info!("VM userspace connection is not successful, Will be placed in crashed state");
                                },
                                Err(response) => {
                                    error!(
                                        "VM: {} has been removed: {response}",
                                        vm.vm_parameters.name
                                    );
                                }
                            }
                        }
                    }
                    if vm_ssr_enablecheck && vm_autostart_done {
                        let vm_instance = instance_for_thread.clone();
                        info!("Auto Shutdown Thread has been initiated");
                        VmInstance::auto_shutdown_thread_handle_initiator(vm_instance);
                    }
                });
                no_fs_dependent_handles.push(handle);
            } else {
                // Wait for timeout if fs dependent
                thread::spawn(move || {
                    let mut vm_ssr_enablecheck:bool = false;
                    let mut watcher = system_properties::PropertyWatcher::new(BOOT_COMPLETE_PROP).unwrap();
                    if let Ok(_) = watcher.wait_for_value(
                        "1",
                        Some(Duration::new(timeout_for_thread.into(), 0)),
                    ) {
                        debug!("System boot completed.");
                        // Only aquire lock after wait for boot complete.
                        // Allows clients to register cbs during the wait.
                        if let Ok(mut vm) = instance_for_thread.lock() {
                            vm.launch_autostart_vm();
                            vm_ssr_enablecheck = vm.vm_parameters.vm_ssr_enable;
                            drop(vm);
                        }
                        if vm_ssr_enablecheck{
                          if let Ok(mut vm) = instance_for_thread.lock() {
                                match vm.autostart_connectvm() {
                                    Ok(-1)=>{
                                        info!("VM userspace connect is not supported");
                                    },
                                    Ok(0)=>{
                                        info!("VM userpspace connection is successful");
                                    },
                                    Ok(_)=>{
                                        info!("VM userspace connection is not successful, Will be placed in crashed state");
                                    },
                                    Err(response) => {
                                        error!(
                                            "Client: {} has been removed: {response}",
                                            vm.vm_parameters.name
                                        );
                                    }
                                }
                            drop(vm);
                            }
                        }
                        if vm_ssr_enablecheck {
                            let vm_instance = instance_for_thread.clone();
                            info!("Auto Shutdown Thread has been initiated");
                            VmInstance::auto_shutdown_thread_handle_initiator(vm_instance);
                        }
                    } else {
                        error!("Timed out checking for sys.boot_completed.");
                        if let Ok(mut vm) = instance_for_thread.lock() {
                            vm.autostart_done = true;
                        }
                    }
                });
            }
        }

        // Join the no_fs_dependent threads
        for handle in no_fs_dependent_handles {
            let _res = handle.join();
        }

        return service;
    }

    fn parse_event(msg: Vec<u8>) -> Option<UeventInfo> {
        if let Ok(msg) = String::from_utf8(msg) {
            let mut pairs: Vec<&str> =
                msg.trim_matches('\0').split('\0').collect();
            pairs.retain(|s| s.contains("="));

            let mut event = None;
            let mut vm_name = None;
            let mut event_reason = None;
            for pair in pairs {
                let p: Vec<&str> = pair.splitn(2, "=").collect();
                match p[0] {
                    "EVENT" => event = Some(p[1].to_string()),
                    "vm_name" => vm_name = Some(p[1].to_string()),
                    "vm_exit" => event_reason = Some(p[1].to_string()),
                    _ => continue,
                };
            }
            let uevent_info = UeventInfo::new(vm_name.clone().unwrap_or_else(|| String::new()),event.clone().unwrap_or_else(|| String::new()),event_reason.clone().and_then(|s| s.parse::<u32>().ok()).unwrap_or(0));
            if !uevent_info.event.is_empty() {
                info!("These are the uevent_info parameter VM Name: {}, \n \t\t UEVENT Received: {}, \n \t\t Uevent Reason: {}", uevent_info.vm_name, uevent_info.event, uevent_info.event_reason.to_string());
            }
            return Some(uevent_info);
        }
        return None;
    }

    fn handle_uevent_event(
        uevent_fd: RawFd,
        vm_instance_map: &Arc<VmInstanceMap>,
    ) -> () {
        let mut msg = Vec::<u8>::new();
        while let Ok(_res) =
            UEvent::uevent_kernel_multicast_recv(uevent_fd, &mut msg, 1024)
        {
            if let Some(event) = Self::parse_event(msg.clone()) {
                match vm_instance_map.get(&event.vm_name) {
                    Some(wrpr) => {
                        if let Ok(mut instance) = wrpr.instance.lock() {
                            debug!("Initiating request to notify state change to {} clients", &event.vm_name);
                            instance.notify_clients(&event.event, &event.event_reason.to_string());
                        }
                    }
                    None => {
                        if !event.vm_name.is_empty() {
                            debug!("Invalid vm_name received from uevent");
                        }
                    }
                }
            }
        }
    }

    fn uevent_listener(vm_instance_map: Arc<VmInstanceMap>) -> () {
        info!("started uevent listener thread");
        let uevent_fd = UEvent::uevent_open_socket(64 * 1024, true).unwrap();
        fcntl(uevent_fd, FcntlArg::F_SETFL(OFlag::O_NONBLOCK)).unwrap();
        let borrowed_uevent_fd = unsafe { BorrowedFd::borrow_raw(uevent_fd) };

        loop {
            let mut ufd = [PollFd::new(borrowed_uevent_fd, PollFlags::POLLIN)];
            if let Ok(nr) = poll(&mut ufd,  PollTimeout::NONE) {
                if nr < 0 {
                    continue;
                }
                if ufd[0].revents().unwrap().contains(PollFlags::POLLIN) {
                    Self::handle_uevent_event(
                        uevent_fd.clone(),
                        &vm_instance_map,
                    );
                }
            }
        }
    }

    // Returns a vector of vmParameters.
    fn parse_vm_config_file() -> Result<Vec<VmParameters>, Box<dyn Error>> {
        let mut vm_parameters_list = Vec::<VmParameters>::new();

        // Uses serde_json to deserialize directly into strongly typed VmParameters object.
        let vendor_config_file = File::open(VENDOR_CONFIG_FILE)?;
        let root: Value = match serde_json::from_reader(vendor_config_file){
            Ok(parsed) => parsed,
            Err(e) => {
                    error!("Parsing of JSON file is incorrect : {}",e);
                    return Err(format!("Error parsing JSON: {}",e).into());
                }
        };
        let json_config_array: &Vec<Value> = root
            .get("qvirtmgr")
            .and_then(|mgr| mgr.get("vm_config"))
            .and_then(|arr| arr.as_array())
            .ok_or("VM Configuration is invalid.")?;
        for config in json_config_array {
            let enable_present = config.get("enable").is_some();
            let enabled_socs_present = config.get("enabled_socs").is_some();
            if enable_present && enabled_socs_present {
                error!("Error: Only one of the predefined keys ('enable', 'enabled_socs') is allowed in config");
                return Err(Box::new(std::io::Error::new(std::io::ErrorKind::Other, "Multiple keys('enable', 'enabled_socs') found")));
            }
            match serde_json::from_value::<VmParameters>(config.to_owned()) {
                Ok(vm_param) => {
                    vm_parameters_list.push(vm_param);
                }
                Err(e) => {
                    // Just skip any malformed vm config.
                    let name: &str = config
                        .get("name")
                        .and_then(|val| val.as_str())
                        .unwrap_or("No Name Specified");
                    error!("Skipping entry for '{}'. Err: {e}", name);
                    continue;
                }
            }
        }
        Ok(vm_parameters_list)
    }
}

impl Interface for VirtualizationService {}

impl IVirtualizationService for VirtualizationService {
    fn getVm(
        &self,
        vm_name: &str,
    ) -> Result<Strong<(dyn IVirtualMachine)>, Status> {
        info!(
            "getVm: Requested vm handle for {} from pid={}",
            vm_name,
            ThreadState::get_calling_pid()
        );

        if let Some(instance_wrpr) = self.vm_instance_map.get(vm_name) {
            if !instance_wrpr.enabled_socs.is_empty() {
                let targets = instance_wrpr.enabled_socs.clone();
                let current_soc = &instance_wrpr.current_soc;
                debug!("Current sku value is {}", current_soc);
                if !targets.contains(current_soc) {
                    error!("getVm: {vm_name} is not enabled on this target, rejecting request.");
                    return Err(Status::new_exception_str(
                        ExceptionCode::UNSUPPORTED_OPERATION,
                        Some("vm not enabled on this target"),
                    ));
                }
            }
            else if !instance_wrpr.enabled {
                error!("getVm: enable bit is false for {vm_name}, rejecting request.");
                return Err(Status::new_exception_str(
                    ExceptionCode::UNSUPPORTED_OPERATION,
                    Some("enable bit false."),
                ));
            }
            // Create a VirtualMachine which wraps the Arc<Mutex<VmInstance>>
            let virtual_machine = VirtualMachine {
                vm_instance: instance_wrpr.instance.to_owned(),
                current_soc: instance_wrpr.current_soc.clone(),
            }; // Increments ref count of instance_obj
            return Ok(virtual_machine.to_binder());
        } else {
            error!("getVm: Invalid name argument passed, rejecting request.");
            return Err(Status::new_exception_str(
                ExceptionCode::ILLEGAL_ARGUMENT,
                Some("Invalid name"),
            ));
        }
    }
}
